sql/expression/function/regexp_instr.go

```diff
@@ -0,0 +1,255 @@
+	if r.Flags != nil {
+		result = append(result, r.Flags)
+	}
+	return result
+}
+
+// Resolved implements the sql.Expression interface.
+func (r *RegexpInstr) Resolved() bool {
+	return r.Text.Resolved() && r.Pattern.Resolved() && r.Position.Resolved() && r.Occurrence.Resolved() &&
+		r.ReturnOption.Resolved() && (r.Flags == nil || r.Flags.Resolved())
+}
+
+// WithChildren implements the sql.Expression interface.
+func (r *RegexpInstr) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	required := 5
+	if r.Flags != nil {
+		required = 6
+	}
+	if len(children) != required {
+		return nil, sql.ErrInvalidChildrenNumber.New(r, len(children), required)
+	}
+	return NewRegexpInstr(children...)
+}
+
+// String implements the sql.Expression interface.
+func (r *RegexpInstr) String() string {
+	var args []string
+	for _, e := range r.Children() {
+		args = append(args, e.String())
+	}
+	return fmt.Sprintf("%s(%s)", r.FunctionName(), strings.Join(args, ","))
+}
+
+// compile handles compilation of the regex.
+func (r *RegexpInstr) compile(ctx *sql.Context) {
+	r.compileOnce.Do(func() {
+		r.re, r.compileErr = compileRegex(ctx, r.Pattern, r.Text, r.Flags, r.FunctionName(), nil)
+	})
+}
+
+// Eval implements the sql.Expression interface.
+func (r *RegexpInstr) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("function.RegexpInstr")
+	defer span.End()
+
+	cached := r.cachedVal.Load()
+	if cached != nil {
+		return cached, nil
+	}
+
+	r.compile(ctx)
+	if r.compileErr != nil {
+		return nil, r.compileErr
+	}
+	if r.re == nil {
+		return nil, nil
+	}
+
+	text, err := r.Text.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if text == nil {
+		return nil, nil
+	}
+	text, _, err = types.LongText.Convert(text)
+	if err != nil {
+		return nil, err
+	}
+
+	pos, err := r.Position.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if pos == nil {
+		return nil, nil
+	}
+	pos, _, err = types.Int32.Convert(pos)
+	if err != nil {
+		return nil, err
+	}
+
+	occurrence, err := r.Occurrence.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if occurrence == nil {
+		return nil, nil
+	}
+	occurrence, _, err = types.Int32.Convert(occurrence)
+	if err != nil {
+		return nil, err
+	}
+
+	returnOption, err := r.ReturnOption.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+	if returnOption == nil {
+		return nil, nil
+	}
+	returnOption, _, err = types.Int32.Convert(returnOption)
+	if err != nil {
+		return nil, err
+	}
+
+	err = r.re.SetMatchString(ctx, text.(string))
+	if err != nil {
+		return nil, err
+	}
+	index, err := r.re.IndexOf(ctx, int(pos.(int32)), int(occurrence.(int32)), returnOption.(int32) == 1)
+	if err != nil {
+		return nil, err
+	}
+
+	outVal := int32(index + 1)
+	if canBeCached(r.Text) {
+		r.cachedVal.Store(outVal)
+	}
+	return outVal, nil
+}
+
+// Close implements the sql.Closer interface.
+func (r *RegexpInstr) Close(ctx *sql.Context) error {
```

Comment:
This looks suspicious to me. AFAIK sql.Expressions get `Dispose()` called on them when the engine is done with the query, but I'm not sure if they get `Close` called. Even if they do, this regexp-related sql.Expression implementations seem to be the only implementations that actually use this. Other things use `Dispose`. We should move them all to `Dispose` in my opinion.